caikit/core/data_model/base.py

```diff
@@ -477,7 +477,20 @@ def __setattr__(self, name, val):
 
         # If attempting to set one of the named fields or a oneof, instead set
         # the private version of the attribute.
-        if name in cls.fields or name in cls._fields_oneofs_map:
+        if name in cls.fields:
+            field_descriptor = cls._proto_class.DESCRIPTOR.fields_by_name[name]
+            # for bytes, convert to bytes if not already
+            if field_descriptor.type == field_descriptor.TYPE_BYTES:
+                error.value_check(
+                    "<COR13072712E>",
+                    isinstance(val, (bytes, str)),
+                    "{} must be string or bytes",
+                    val,
+                )
+                if isinstance(val, str):
+                    val = val.encode("utf-8")
+            super().__setattr__(f"_{name}", val)
```

Comment:
I'm curious why we aren't checking other field types while doing the `__setattr__` 🤔

Reply:
Hmm. This is an interesting. It seems like the logical place to introduce type casting, but that also feels like it could be a pretty big can of worms to open with the potential for a major performance hit to run through all of the `if isinstance(...)` checks. The alternative would be to simply reject attempts to set fields with the wrong type. I think I might already have something like this written somewhere...

Reply:
So in that case, we would reject `f2 = Foo("asdf")` for `foo: bytes`?

Reply:
Did you mean the `checker = proto_type_checkers.GetTypeChecker(field_descriptor)` code for checking the type?

Reply:
Yes! That's what I was trying to remember

Reply:
This idea starts complaining about things like this now:
```
@dataobject
    class Foo(DataObjectBase):
        foo: str
        bar: int

    inst = Foo()
```
I'm guessing introducing type-checking within `setattr` is going to break a bunch of stuff